sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -109,6 +109,32 @@ class VdmSubtypeIndex(Enum):
         "vdm_supported": "N/A"
         }
 
+def read_only_cache_none_dict_and_list(func):
+    cache_name = f'_{func.__name__}_cache'
+    def wrapper(self):
+        if not hasattr(self, cache_name):
+            cache_value = func(self)
+            setattr(self, cache_name, cache_value)
+        else:
+            cache_value = getattr(self, cache_name)
+            if cache_value is None:
+                cache_value = func(self)
+                setattr(self, cache_name, cache_value)
+        return cache_value
+    return wrapper
+
+# Decorator that caches return values of dict/list methods only when non-empty collections are returned
+def read_only_cache_dict_and_list(func):
+    cache_name = f'_{func.__name__}_cache'
+    def wrapper(self):
+        if not hasattr(self, cache_name):
+            cache_value = func(self)
+            if cache_value is not None and isinstance(cache_value, (dict, list)) and len(cache_value) > 0:
+                setattr(self, cache_name, cache_value)
+            return cache_value
+        return getattr(self, cache_name)
+    return wrapper
```

Comment:
@jianyuewu can we have only one function implementation for cached value be it return value from collections or otherwise? i.e just need `read_only_cached_api_return()` ?

Reply:
Combined cache function will have more complicated logic, I can try it.